### PR TITLE
Add Rayon support

### LINKED ISSUE #16: Rayon support
It'd be great to have Rayon compatibility for hashbrown's `HashMap` and `HashSet`. I think it's a case of implementing `IntoParallelIterator`, `FromParallelIterator` and `ParallelExtend`.

### PR DESCRIPTION
**Pull request by Amanieu**
Taking over from #30 since the author seems to be inactive.

Fixes #16 

cc @hcpl @cuviper

### CODE REVIEW

**Comment by hcpl** [comment]
I apologise for not responding. After reading review comments in my PR, I thought that making proposed changes will require several hours of debugging segfaults (like the pointer range calculation bug), which I didn't feel like doing at the time. Re-reading comments and skimming your https://github.com/Amanieu/hashbrown/pull/37/commits/0f8d03539184f56e58ca8b711af9a5d0079c7128 made me realise that it's easy enough to deal with in an hour or so, so I overestimated the time needed.

That said, thanks for taking over and making progress!

**Comment by Amanieu** [comment]
It's fine, I just wanted to get this in so that the `RawIter` changes can be ported to the libstd PR as well. I want to keep libstd and hashbrown in sync as much as possible.

Sorry about the `split` bug, this was my fault for being careless.

**Comment by cuviper** [comment]
> I want to keep libstd and hashbrown in sync as much as possible.

Are you going to keep maintaining hashbrown (w/ rayon) once it's in std?  This would make `rayon-hash` obsolete, which is fine by me.  Or if std exposes enough of the `RawIter` interface to stable users, then `rayon` itself could steal the parallel implementation from here.

**Comment by Amanieu** [comment]
I will keep maintaining it for a while, but I eventually plan on making some breaking changes to the `HashMap` API, at which point I will change the version to 0.2.

**Comment by cuviper** [comment]
OK, then maybe I'll keep going with rayon-hash too, keeping std-compatible.  It's only fair to steal the code back when AFAICT the impl here started as a copy of rayon-hash's. :smile:

File: src/external_trait_impls/rayon/raw.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        folder.consume_iter(self.iter)
+    }
+}
+
+/// Parallel iterator which consumes a table and returns elements.
+pub struct RawIntoParIter<T> {
+    iter: RawIterRange<T>,
+    alloc: Option<(NonNull<u8>, Layout)>,
+}
```
Comment (cuviper) [inline src/external_trait_impls/rayon/raw.rs]: If this struct gets dropped without ever driving it, then I think you'll leak both the items and the raw allocation.

File: src/external_trait_impls/rayon/raw.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        });
+        let producer = ParDrainProducer { iter: self.iter };
+        plumbing::bridge_unindexed(producer, consumer)
+    }
+}
+
+/// Parallel iterator which consumes elements without freeing the table storage.
+pub struct RawParDrain<'a, T> {
+    table: &'a mut RawTable<T>,
+}
```
Comment (cuviper) [inline src/external_trait_impls/rayon/raw.rs]: If this struct gets dropped without ever driving it, then it won't have drained any items.

**Comment by Amanieu** [comment]
The leak is fixed by #38.

**Pull request merged**